kernel/msg_q.c

```diff
@@ -319,13 +319,23 @@ int z_impl_k_msgq_peek_more(struct k_msgq *msgq, void *data, uint32_t idx)
 {
 	k_spinlock_key_t key;
 	int result;
+	uint32_t items_leaving;
 
 	key = k_spin_lock(&msgq->lock);
 
 	if (msgq->used_msgs > idx) {
-		/* take first available message from queue */
-		(void)memcpy(data, msgq->read_ptr + (idx * msgq->msg_size), msgq->msg_size);
-		result = 0;
+		/* check item available in start/end of ring buffer */
+		if (msgq->read_ptr + (idx * msgq->msg_size) >= msgq->buffer_end) {
+			items_leaving = (msgq->buffer_end - msgq->read_ptr) / msgq->msg_size;
+			(void)memcpy(data,
+				msgq->buffer_start + ((idx - items_leaving) * msgq->msg_size),
+				msgq->msg_size);
+			result = 0;
+		} else {
+			/* take available message from queue */
+			(void)memcpy(data, msgq->read_ptr + (idx * msgq->msg_size), msgq->msg_size);
+			result = 0;
+		}
```

Comment:
It would be better to avoid a division if possible as they can be very costly operations--particularly for architectures that do not have built in division instruction. As such, I would recommend converting the logic slightly to work on a byte offset basis.  For example (my apologies if I have the logic slightly wrong--I wrote it only to help convey a couple of ideas):

```
bytes_to_end = (msgq->buffer_end - msgq->read_ptr);   /* Cache # of bytes to the end of msgq buffer */

byte_offset = idx * msgq->msg_size;                   /* Calculate offset and starting address */
start_addr = msgq->read_ptr;                          /* for the simplest success case */

if (bytes_to_end <= byte_offset) {
    byte_offset -= bytes_to_end;            /* Tweak the values in case */
    start_addr = msgq->buffer_start;        /* wrap-around is required */
}

(void) memcpy(data, start_addr + byte_offset, msgq->msg_size);
```